src/MvvmNano.Forms/MvvmNanoFormsPresenter.cs

```diff
@@ -90,21 +92,39 @@ public void SetApplication(Application application)
                 .ToArray();
         }
 
+        bool IsMasterDetailApplication()
+        {
+            return Application is MvvmNanoMasterDetailApplication;
+        }
+
+        public void ChangeRootViewModel<TViewModel>() where TViewModel : MvvmNanoViewModel
+        {
+            var viewName = GetViewNameByViewModel(typeof(TViewModel));
+            Type viewType = GetViewTypeByName(viewName); 
+            if(IsMasterDetailApplication() && viewType.GetTypeInfo().IsSubclassOf(typeof(NanoMasterDetailPage)))
+                ((MvvmNanoMasterDetailApplication)Application).SetUpMasterDetailPage<TViewModel>();    
+            else
+                Application.SetUpMainPage<TViewModel>();
+        } 
+
+        public string GetViewNameByViewModel(Type viewModelType)
+        {
+            return viewModelType.Name.Replace(VIEW_MODEL_SUFFIX, VIEW_SUFFIX);
+        } 
+
         /// <summary>
         /// Navigates to a Page and automatically creates a new instance of
         /// the corresponding View Model. Also passes some parameters of the
         /// given type.
         /// </summary>
         public void NavigateToViewModel<TViewModel, TNavigationParameter>(TNavigationParameter parameter)
         {
-            var viewModel = CreateViewModel<TViewModel, TNavigationParameter>();
+            IViewModel<TNavigationParameter> viewModel = CreateViewModel<TViewModel, TNavigationParameter>();
             viewModel.Initialize(parameter);
-
-            IView view = CreateViewFor<TViewModel>();
-            view.SetViewModel(viewModel);
-
-            OpenPage(view as Page);
-        }
+            IView viewFor = CreateViewFor<TViewModel>();
+            viewFor.SetViewModel((IViewModel)viewModel);
+            StartOpeningPage<TViewModel>(viewFor as Page);
```

Comment:
I guess this will cause some headaches. Because this `NavigateToViewModel()` method cannot be awaited (void), and the execution of `StartOpeningPage<TViewModel>()` _is_ not be awaited. So whenever a caller of `NavigateToViewModel()` is doing something right after it, it may be disturbed by the then finishing  `StartOpeningPage<TViewModel>()`.

I am not 100% sure about deadlocks, but I guess `StartOpeningPage<TViewModel>(viewFor as Page).GetAwaiter().GetResult()` should be a compromise.